File: source-stl/listmakhluk/HimpMakhluk.cpp
/** Implementasi Kelas Himpunan Makhluk yang mengatur kumpulan makhluk (versi STL - vector)
 * Kontributor : BLP (13514047)
 * */

#include "HimpMakhluk.h"

using namespace std;

/** Inisialisasi Konstanta MaxMakhluk dengan 1 */
const unsigned int HimpMakhluk::MaxMakhluk = 1;

HimpMakhluk::HimpMakhluk() {
  //kosong tidak perlu alokasi dahulu
  //alokasi hanya saat Add Makhluk
}

HimpMakhluk::~HimpMakhluk() {
  /** \var ~HimpMakhluk()
  * \brief Destruktor himpunan makhluk akan menghapus memori yang telah dialokasikan
  * */
  if (!Mem.empty()) {
  	Mem.clear();
  }
}

void HimpMakhluk::AddMakhluk(Makhluk*& A) {
  /** \var void AddMakhluk(Makhluk*& A)
   * \brief Menambahkan petunjuk ke makhluk ke dalam memori
   * \param A sebagai input petunjuk dari makhluk
   *  */
  Mem.push_back(A);
}

Makhluk *  HimpMakhluk::SurvFight() {
	/** \var void SurvFight()
	 * \brief Sebagai eksekusi saat mencapai kondisi untuk bertengkar
	 * */
	//Prekondisi makhluk dalam himpunan telah lebih dari 2
	vector<Makhluk *>::iterator pl1 = Mem.begin();
	vector<Makhluk *>::iterator pl2 = pl1++;
	Makhluk * M;
	if ((*pl1)->getpowerlevel() > (*pl2)->getpowerlevel())
	{
		M = DeleteMakhluk(*pl2);
	} else {
		M = DeleteMakhluk(*pl1);
	}

	return M;
}



Makhluk* HimpMakhluk::DeleteMakhluk(Makhluk*& A) {
  /** \var Makhluk* DeleteMakhluk(Makhluk*& A)
   * \brief Berguna untuk menghapus elemen atau makhluk yang telah kalah.
   * Kembalian sebagai petunjuk makhluk agar dapat diketahui makhluk yang dihapus.
   * Kembalian NULL jika tidak ada yang dihapus
   * Kembalian petunjuk makhluk yang dihapus jika berhasil dihapus.
   * \param A Sebagai indikator makhluk yang akan dihapus.
   * */
   //Inisialisasi yang akan dihapus
   //Menggunakan algoritma pencarian traversal anggota list
  Makhluk* DelMakhluk = NULL;
  if (!Mem.empty())
  {
    bool found = false;
    vector<Makhluk *>::iterator it = Mem.begin();
    do {
      if ((*it)->gettype() == A->gettype()) //cek identitas
      {
        DelMakhluk = (*it);
        Mem.erase(it);
        found = true;
      } else
      {
        it++;
      }
      
	} while ((!found)&&(it != Mem.end()));
  }
  return DelMakhluk;
}

bool HimpMakhluk::isFull() {
  /** \var bool IsFull()
   * \brief Mengecek apakah sudah mencapai kondisi pertarungan
   * */
  if (Mem.size()==MaxMakhluk) {
    return true;
  } else {
    return false;
  }
}

vector<Makhluk *>::iterator HimpMakhluk::GetFirst(){
	return Mem.begin();
}

vector<Makhluk *>::iterator HimpMakhluk::GetLast(){
	return Mem.end();
}

unsigned int HimpMakhluk::GetnMakhluk(){
	return Mem.size();
}
unsigned int HimpMakhluk::GetMaxMakhluk(){
	return MaxMakhluk;
}

bool HimpMakhluk::isOver(){
	if (Mem.size() > MaxMakhluk){
		return true;
	}
	else{
		return false;
	}
}

bool HimpMakhluk::isEmpty(){
	return Mem.empty();
}


Makhluk* HimpMakhluk::checkMoveMakhluk(int x, int y)
{
	/*! \var Node* checkMoveMakhluk
	*  \brief Mengecek apakah makhluk didalam sudah pindah atau belum kalau sudah dipindahkan
	*  \param x x adalah parameter absis list pada matriks
	*  \param y y adalah parameter ordinat list pada matriks
	*/
	Makhluk *M = NULL;
	bool end = false;
	int Px;
	int Py;
	vector<Makhluk *>::iterator it = Mem.begin();
	int i = 1;
	int size = Mem.size();
	while ((i<=size) && (end == false))
	{
		Px = (* it)->getX();
		Py = (* it)->getY();

		if (((Px != x) || (Py != y)))
		{
			M = DeleteMakhluk((*it));
			end = true;
		}
		else
		{
			it++;
			i++;
		}
	}

	return M;
}

File: src/java/HimpMakhluk.java
import java.util.*;

/**
 * @author Joshua Salimin - 13514001
 */
public class HimpMakhluk {

    /**
     * MaxMakhluk sebagai batas makhluk yang diperbolehkan dalam himpunan makhluk
     * al sebagai penyimpan himpunan makhluk
     */
    private final int MaxMakhluk= 1;
    private List<Makhluk> al;
    
    /**
     * Konstruktor kelas HimpMakhluk
     */
    public HimpMakhluk(){
        al = new ArrayList<Makhluk>();
    }
    
    /**
     * Sebuah prosedur untuk menambahkan sebuah Makhluk ke dalam ArrayList
     * @param M merupakan Makhluk yang akan ditambahkan ke dalam ArrayList
     */
    public void AddMakhluk(Makhluk M){
        al.add(M);
    }
    
    /**
     * Sebuah fungsi yang mengembalikan Makhluk yang dihapus dari ArrayList dimana Makhluk tersebut memiliki powerlevel yang lebih kecil
     * @return Makhluk merupakan Makhluk yang dihapus dari ArrayList
     */
    public Makhluk SurvFight(){
        /*Makhluk M = null;
        if(al.get(0).powerlevel > al.get(1).powerlevel) {
			M = DeleteMakhluk(al.get(1));
		} else {
			M = DeleteMakhluk(al.get(1));
		}
		return M;*/
        
        int temppower=al.get(0).powerlevel;
        int id=al.get(0).type;
		int idxfound = 0;
        for(int j=0;j<al.size();++j){
            if(temppower<al.get(j).powerlevel){
                temppower=al.get(j).powerlevel;
                idxfound = j;
            }
        }
		
		return DeleteMakhluk(al.get(idxfound));
		
    }
    
    /**
     * Sebuah fungsi yang akan mencari Makhluk dan menghapus Makhluk tersebut dari ArrayList
     * @param M merupakan masukan sebuah Makhluk yang akan dihapus
     * @return Makhluk merupakan makhluk yang dihapus
     */
    public Makhluk DeleteMakhluk(Makhluk M){
        boolean found  = false;
		if(al.size()!=0){
            int i =0;
            while(i<al.size() && !found){
                if(al.get(i).type == M.type )
                    found = true;
                 else
                    ++i;
            }
			if(found) {
				Makhluk S = al.get(i);
				al.remove(i); 
				return S;
			}
			else {
				return null;
			}
        }
		else {
			return null;
		}
    }   
    
    /**
     *sebuah fungsi yang mengembalikan boolean apakah ArrayList penuh atau tidak
     * @return true/false boolean apakah true atau false
     */
    public boolean isFull(){
        if(al.size()==MaxMakhluk){
            return true;
        }
        else{
            return false;
        }
    }
    
    /**
     *sebuah fungsi yang mengembalikan Makhluk di indeks ke i
     * @param i merupakan posisi indeks 
     * @return al.get(i) Makhluk yang merupakan di cari
     */
    public Makhluk GetMakhlukin(int i) {
		return al.get(i);
	}
	
    /**
     * Sebuah fungsi yang mengembalikan integer, banyaknya Makhluk dalam ArrayList
     * @return al.size() size daripada ArrayList
     */
    public int GetnMakhluk(){
        return al.size();
    }
    
    /**
     * Sebuah fungsi yang mengembalikan integer yaitu banyaknya Maksimum Makhluk yang dapat ditampung dalam satu ArrayList
     * @return MaxMakhluk
     */
    public int GetMaxMakhluk(){
        return MaxMakhluk;
    }
    
    /**
     * Sebuah fungsi yang mengembalikan boolean apakah size dari ArrayList lebih besar dari Maksimum Makhluk yang dapat ditampungg
     * @return true/false 
     */
    public boolean isOver(){
        return al.size()>MaxMakhluk;
    }
    
    /**
     * Sebuah fungsi yang mengembalikan boolean dan mengecek apakah ArrayList kosong atau tidak
     * @return true/false
     */
    public boolean isEmpty(){
        return al.size()==0;
    }
    
    /**
     * Sebuah fungsi yang mengembalikan Makhluk dan menghapus Makhluk dari ArrayList jika Makhluk tersebut sudah pindah
     * @param x absis posisi dari Makhluk
     * @param y ordinat posisi dari Makhluk
     * @return Makhluk yang sudah berpindah
     */
    public Makhluk checkMoveMakhluk(int x, int y){
        //dianggap dalam suatu array hanya terdapat 1 makhluk
        Makhluk M = null;
        int i=0;
        int px,py;
        boolean end =false;
        while(i<al.size() && !end){
            px = al.get(i).x;
            py = al.get(i).y;
            if(px!=x || py!=y){
                 M = DeleteMakhluk(al.get(i));
                end = true;
            }
            else{
              ++i;  
            } 
        }
        
        return M;
    }
    
}
